compiler/rustc_passes/src/dead.rs

```diff
@@ -236,6 +236,38 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
         }
     }
 
+    fn handle_offset_of(&mut self, expr: &'tcx hir::Expr<'tcx>) {
+        let data = self.typeck_results().offset_of_data();
+        let &(container, ref indices) =
+            data.get(expr.hir_id).expect("no offset_of_data for offset_of");
+
+        let mut last_did = expr.hir_id.owner.to_def_id();
+        let mut current_ty = container;
+
+        for &index in indices {
+            match current_ty.kind() {
+                ty::Adt(def, subst) => {
+                    let field = &def.non_enum_variant().fields[index];
+
+                    self.insert_def_id(field.did);
+                    let field_ty = field.ty(self.tcx, subst);
+
+                    last_did = field.did;
+                    current_ty =
+                        self.tcx.normalize_erasing_regions(self.tcx.param_env(field.did), field_ty);
```

Comment:
The param_env is wrong here. This should use the same param_env as typeck_results, because that's in that param_env that `subst` are expressed.

Reply:
I'm not really sure how to get the right DefId for the param_env in this location.

Reply:
Since @DrMeepster asked for a little help on this I made a counterexample that currently ICEs:
```rust
#![feature(offset_of)]

// This struct contains a projection that can only be normalized after getting the field type.
struct A<T: Project> {
    a: <T as Project>::EquateParamTo,
}

// This is the inner struct that we want to get.
struct MyFieldIsNotDead {
    not_dead: u8,
}

// These are some helpers.
// Inside the param env of `test`, we want to make it so that it considers T=MyFieldIsNotDead.
struct GenericIsEqual<T>(T);
trait Project {
    type EquateParamTo;
}
impl<T> Project for GenericIsEqual<T> {
    type EquateParamTo = T;
}

fn test<T>() -> usize
where
    GenericIsEqual<T>: Project<EquateParamTo = MyFieldIsNotDead>,
{
	// The first field of the A that we construct here is `<GenericIsEqual<T>> as Project>::EquateParamTo`.
	// Typeck normalizes this and figures that the not_dead field is totally fine and accessible.
	// But importantly, the normalization ends up with T, which, as we've declared in our param env is MyFieldDead.
	// When we're in the param env of the `a` field, the where bound above is not in scope, so we don't know what T is - it's generic.
	// We cannot access a field on T. Boom!
    std::mem::offset_of!(A<GenericIsEqual<T>>, a.not_dead)
}

fn main() {
    test::<MyFieldIsNotDead>();
}
```
```
error: internal compiler error: compiler/rustc_passes/src/dead.rs:269:22: named field access on non-ADT
  --> uwu.rs:27:5
   |
27 |     std::mem::offset_of!(A<GenericIsEqual<T>>, a.not_dead)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
Would `cx.typeck_results().hir_owner.to_def_id()` work?